### PR TITLE
WIP: Load IC states from other tools

### PR DESCRIPTION
**Pull request by sanguinariojoe**
Add the capability to load an Initial Condition exported with a quasi-static solver. For the time being it is just tested on trivial examples.

It should still be documented.

~~This PR is combined with PRs on some other projects:~~

 ~~- [X] MoorPy: https://github.com/NREL/MoorPy/pull/30~~
 ~~- [ ] Map++:~~

### CODE REVIEW

**Review by RyanDavies19 — APPROVED** [review]
Looks good to me, I'll admit a good part of this is over my head but it seems like it is all in handling I/O via coupling with other codes so no major changes to internal workings or existing use methods. Cool to see the integration with MoorPy too!

**Comment by RyanDavies19** [comment]
The only thing I'll add @sanguinariojoe is that we might want to make a note somewhere in the code via a comment that changes/updates to the MoorDyn states will require changes here. Notably for both ongoing work on viscoelastic modeling and VIV additional state variables will need to be added.

**Comment by sanguinariojoe** [comment]
> The only thing I'll add @sanguinariojoe is that we might want to make a note somewhere in the code via a comment that changes/updates to the MoorDyn states will require changes here. Notably for both ongoing work on viscoelastic modeling and VIV additional state variables will need to be added.

I do not know how you are planning to do that, so just point me where you want that note and I will be more than happy to add it.

This is a bit frozen until i get some feedback from MoorPy guys

**Comment by RyanDavies19** [comment]
> > The only thing I'll add @sanguinariojoe is that we might want to make a note somewhere in the code via a comment that changes/updates to the MoorDyn states will require changes here. Notably for both ongoing work on viscoelastic modeling and VIV additional state variables will need to be added.
> 
> I do not know how you are planning to do that, so just point me where you want that note and I will be more than happy to add it.
> 
> This is a bit frozen until i get some feedback from MoorPy guys

A note around here would be good, because that's where new states will be added: https://github.com/FloatingArrayDesign/MoorDyn/blob/60d9dddc80f145029baeadf936a1343634dd8c2f/source/State.hpp#L260

I chatted with @mattEhall about the PR to the MoorPy repo yesterday. Ideally it would be best to keep MoorPy separate from MoorDyn as it is typically coupled with frequency domain analysis tools like [RAFT](https://github.com/WISDEM/RAFT). However the idea of being able to export a MoorPy system to MoorDyn still seems like a super useful idea, especially for optimization. Would it be possible for this to live as a function in the MoorDyn python wrapper instead of a change to MoorPy? It would take a MoorPy system object as it's input and from there export it to MoorDyn states.

**Comment by sanguinariojoe** [comment]
> A note around here would be good, because that's where new states will be added: 

https://github.com/FloatingArrayDesign/MoorDyn/pull/244/commits/a3dc662cac02778af544427c251c1924b5b50c59

> Would it be possible for this to live as a function in the MoorDyn python wrapper instead of a change to MoorPy? It would take a MoorPy system object as it's input and from there export it to MoorDyn states.

https://github.com/FloatingArrayDesign/MoorDyn/pull/244/commits/425ac7fa39bdd8de789319eea04516642d35f335

> I also had a more general question about the wrappers: Do the wrappers automatically update when changes are made to MoorDyn2.h or do those changes also need to be made in the wrapper code? I.e. to what degree do you touch the wrapper code when updating the MoorDyn API?

Nope, I have to manually include them. For Python something can be done, but it will add complexity to the base code. For Fortran and Matlab I do not know a way we can avoid doing it manually. For Rust everything is automatic

---

 - I am fixing some issues with VTK (https://github.com/sanguinariojoe/vtk-builds/actions/runs/10414799968). Until I am not able to run that pipeline, the CI checks will be failing.
 - I am still playing a bit with MoorPy. After I have something more solid I will add some changes here, as well as some documentation on how users can create and use their own initial conditions with MoorPy

**Comment by RyanDavies19** [comment]
> 
> * I am still playing a bit with MoorPy. After I have something more solid I will add some changes here, as well as some documentation on how users can create and use their own initial conditions with MoorPy

If the MoorPy export lived in the MoorDyn python package I think it would work out nicely regardless of the MoorDyn wrapper you use. Embedding it in the MoorDyn wrapper would let users put it in-line with whatever MoorPy driver code they are using. I am envisioning a workflow like this:

```
import moorpy as mp
import moordyn

# ----- set up the mooring system and floating body -----

# Create new MoorPy System and set its depth
mp_system = mp.System(file='lines.txt')

# ----- run the model to demonstrate -----

mp_system.initialize()                                             # make sure everything's connected

# find equilibrium of just the mooring lines (the body is 'coupled' and will by default not be moved)
mp_system.solveEquilibrium()                                       # equilibrate

moordyn.export_mp_state(mp_system, outFileName, version_major=2, version_minor=4)
```

**Comment by sanguinariojoe** [comment]
> I am envisioning a workflow like this:

That is what I implemented, as well as a wrapper on top of that:

https://github.com/core-marine-dev/MoorDyn/blob/stationary/wrappers/python/moordyn/moorpyic.py#L106

Once the python wrapper is installed, it can be used exactly as you posted (except the function is called `moordyn.moorpy_export` instead of `moordyn.export_mp_state`). I can change the function name if you like yours more.

Anyway, I created a PR on MoorPy to give support to Rods (as phony lines). I would get that merged (or rejected) before continuing this work, because such PR would imply some minor changes here.

**Comment by RyanDavies19** [comment]
> > > I am envisioning a workflow like this:
> > 
> > 
> > That is what I implemented, as well as a wrapper on top of that:
> > https://github.com/core-marine-dev/MoorDyn/blob/stationary/wrappers/python/moordyn/moorpyic.py#L106
> > Once the python wrapper is installed, it can be used exactly as you posted (except the function is called `moordyn.moorpy_export` instead of `moordyn.export_mp_state`). I can change the function name if you like yours more.
> > Anyway, I created a PR on MoorPy to give support to Rods (as phony lines). I would get that merged (or rejected) before continuing this work, because such PR would imply some minor changes here.
> 
> Thanks for that, it looks solid. I like your function name even more, it's more explicit. With regards to the MoorPy PR the team is meeting to discuss that soon. I know that there is a desire to keep MoorDyn and MoorPy as structurally similar as possible, so there might be an effort soon to add rod objects.

@sanguinariojoe I talked with the MoorPy team and they are looking to add a Rod type object to MoorPy in the near future and for now are going to leave the code as is. I still think this is a useful tool here that can be merged, we can just caveat that it is not currently compatible with systems that have rods.

**Comment by sanguinariojoe** [comment]
Yeah, this week I have been quite busy. Next week I am adding docs to this
and merging

**Comment by RyanDavies19** [comment]
@sanguinariojoe those docs are great, the python example is super useful for showing how this would all work

File: docs/initialization.rst [hunk 1]
```diff
@@ -0,0 +1,330 @@
+The system initialization, i.e., the generation of the Initial Condition (IC), is of paramount importance, as it can significantly impact the performance and accuracy of the simulation. Several methods are offered in MoorDyn-C version 2, which can be concatenated to achieve the required accuracy at a minimum cost.
```
Comment (RyanDavies19) [inline docs/initialization.rst:1]: In a future PR let's remove these lines (1-5). Lines 1-5 are duplicates of 14-18

**Comment by sanguinariojoe** [comment]
No prob, i'll do that! I am planning to work a little bit more on the
initialization before moving on to time integration again

**Pull request merged**